fpga_interchange/pseudo_pip_model.cc

```diff
@@ -0,0 +1,475 @@
+    std::vector<int32_t> &unused_pseudo_pips = scratch;
+    unused_pseudo_pips.clear();
+    unused_pseudo_pips.reserve(pseudo_pips_for_site.size());
+
+    HashTables::HashMap<int32_t, PseudoPipBel> used_bels;
+    for(int32_t pseudo_pip : pseudo_pips_for_site) {
+        if(!active_pseudo_pips.count(pseudo_pip)) {
+            unused_pseudo_pips.push_back(pseudo_pip);
+            continue;
+        }
+
+        PipId pip;
+        pip.tile = tile;
+        pip.index = pseudo_pip;
+        for(const PseudoPipBel & bel: ctx->pseudo_pip_data.get_logic_bels_for_pip(ctx, site, pip)) {
+            used_bels.emplace(bel.bel_index, bel);
+        }
+    }
+
+    if(unused_pseudo_pips.empty()) {
+        return;
+    }
+
+    int32_t tile_type = ctx->chip_info->tiles[tile].type;
+    const TileTypeInfoPOD & type_data = ctx->chip_info->tile_types[tile_type];
+
+    // This section builds up LUT mapping logic to determine which LUT wires
+    // are availble and which are not.
+    const std::vector<LutElement> &lut_elements = ctx->lut_elements.at(tile_type);
+    std::vector<LutMapper> lut_mappers;
+    lut_mappers.reserve(lut_elements.size());
+    for (size_t i = 0; i < lut_elements.size(); ++i) {
+        lut_mappers.push_back(LutMapper(lut_elements[i]));
+    }
+
+    const TileStatus & tile_status = ctx->tileStatus.at(tile);
+    for (CellInfo *cell : tile_status.sites[site].cells_in_site) {
+        if (cell->lut_cell.pins.empty()) {
+            continue;
+        }
+
+        BelId bel = cell->bel;
+        const auto &bel_data = bel_info(ctx->chip_info, bel);
+        if (bel_data.lut_element != -1) {
+            lut_mappers[bel_data.lut_element].cells.push_back(cell);
+        }
+    }
+
+    std::vector<CellInfo> lut_cells;
+    lut_cells.reserve(used_bels.size());
+    for(const auto & bel_pair : used_bels) {
+        const PseudoPipBel &bel = bel_pair.second;
+        const BelInfoPOD & bel_data = type_data.bel_data[bel.bel_index];
+
+        // This used BEL isn't a LUT, skip it!
+        if(bel_data.lut_element == -1) {
+            continue;
+        }
+
+        lut_cells.emplace_back();
+        CellInfo &cell = lut_cells.back();
+
+        cell.bel.tile = tile;
+        cell.bel.index = bel_pair.first;
+
+        if(ctx->wire_lut == nullptr) {
+            continue;
+        }
+
+        cell.type = IdString(ctx->wire_lut->cell);
+        NPNR_ASSERT(ctx->wire_lut->input_pins.size() == 1);
+        cell.lut_cell.pins.push_back(IdString(ctx->wire_lut->input_pins[0]));
+        cell.lut_cell.equation.resize(2);
+        cell.lut_cell.equation.set(0, false);
+        cell.lut_cell.equation.set(1, true);
+
+        // Map LUT input to input wire used by pseudo pip.
+        IdString input_bel_pin(bel_data.ports[bel.input_bel_pin]);
+        cell.cell_bel_pins[IdString(ctx->wire_lut->input_pins[0])].push_back(input_bel_pin);
+
+        lut_mappers[bel_data.lut_element].cells.push_back(&cell);
```

Comment:
this seems fragile at a glance, couldn't subsequent `emplace_back`s invalidate `&cell`?

Reply:
never mind, I missed the call to `reserve`